DRA kubelet: add dra_resource_claims_in_use gauge vector

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The new metric informs admins whether DRA in general (special "driver_name: <any>" label) and/or specific DRA drivers (other label values) are in use on nodes. This is useful to know because removing a driver is only safe if it is not in  use. If a driver gets removed while it has prepared a ResourceClaim, unpreparing that ResourceClaim and stopping pods is blocked.
    
The implementation of the metric uses read locking of the claim info cache. It retrieves "claims in use" and turns those into the metric.
    
The same code is also used to log changes in the claim info cache with a diff. This hooks into a write update of the claim info cache and uses contextual logging.

#### Which issue(s) this PR fixes:

Slack discussion: https://kubernetes.slack.com/archives/C0409NGC1TK/p1746168044475379?thread_ts=1746001550.655339&cid=C0409NGC1TK
Related-to: https://github.com/kubernetes/enhancements/issues/4381 (GA?)

#### Special notes for your reviewer:

The unit tests check that metrics get calculated. The e2e_node test checks that kubelet really exports the metrics data.

While at it, some bugs in the claiminfo_test.go get fixed: the way how the cache got populated in the test did not match the code anymore.

Let's review this proposal, then document it as part of the 1.34 KEP update before merging the implementation.

/hold
/assign @bart0sh 

#### Does this PR introduce a user-facing change?
```release-note
The new `dra_resource_claims_in_use` kubelet metrics informs about active ResourceClaims, overall and by driver.
```

pohly committed:
DRA kubelet: add dra_resource_claims_in_use gauge vector

The new metric informs admins whether DRA in general (special "driver_name: <any>"
label) and/or specific DRA drivers (other label values) are in use on nodes.
This is useful to know because removing a driver is only safe if it is not in
use. If a driver gets removed while it has prepared a ResourceClaim,
unpreparing that ResourceClaim and stopping pods is blocked.

The implementation of the metric uses read locking of the claim
info cache. It retrieves "claims in use" and turns those into the metric.

The same code is also used to log changes in the claim info cache with
a diff. This hooks into a write update of the claim info cache and uses
contextual logging.

The unit tests check that metrics get calculated. The e2e_node test checks that
kubelet really exports the metrics data.

While at it, some bugs in the claiminfo_test.go get fixed: the way how the
cache got populated in the test did not match the code anymore.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -656,6 +695,10 @@ func createTestObjects(ctx context.Context, clientSet kubernetes.Interface, node
 			RestartPolicy: v1.RestartPolicyNever,
 		},
 	}
+	if podName == "pausepod" {
```
@bart0sh: I considered adding another parameter to `createTestObjects` but decided against it because it would have affected all other call sites. Up for debate...

bart0sh replied:
It looks ok to me, considering the alternative.

pohly replied:
So let's keep it. I added tests for it.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -382,6 +384,43 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 			gomega.Eventually(kubeletPlugin2.GetGRPCCalls).WithTimeout(retryTestTimeout).Should(testdriver.NodeUnprepareResourcesSucceeded)
 		})
 
+		ginkgo.It("must provide metrics", func(ctx context.Context) {
```
This is an e2e_node test because it is easier to get the kubelet metrics there.

If that could be done also in an E2E test, then putting the test there would be more appropriate. The test doesn't really depend on the kubelet configuration and E2E tests are easier to run.

bart0sh commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -142,9 +166,42 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+		changed := false
+		for _, inUse := range delta {
+			switch {
+			case inUse.Count == 0 && inUse.DriverName != "":
```
I'd vote for `<any>`. it looks much more explicit than "" to me.

ArangoGutierrez commented on pkg/kubelet/cm/dra/metrics_test.go:
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2024 The Kubernetes Authors.
```
```suggestion
Copyright 2025 The Kubernetes Authors.
```

ArangoGutierrez commented on test/e2e_node/dra_test.go:
```diff
@@ -382,6 +384,43 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+
+			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "pausepod", false, []string{kubeletPlugin1Name, kubeletPlugin2Name})
+
+			ginkgo.By("wait for pod to succeed")
```
```suggestion
			ginkgo.By("Waiting for pod to succeed")
```
As recommended at https://onsi.github.io/ginkgo/#documenting-complex-specs-by

pohly replied:
You are right, but `dra_test.go` already deviates from that recommendation. To ensure consistency, I would have to update the complete file - I prefer not to do that and instead use the same style also in this PR.

@bart0sh: your preference?

bart0sh replied:
+1 for keeping the same style

ArangoGutierrez commented on test/e2e_node/dra_test.go:
```diff
@@ -382,6 +384,43 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+				}),
+			}), "metrics while pod is running")
+
+			ginkgo.By("delete pod")
```
```suggestion
			ginkgo.By("Deleting pod")
```

ArangoGutierrez reviewed (commented):
Non blocking comments

pohly commented:
e2e-kind failed because of https://github.com/kubernetes/kubernetes/issues/131748.

bart0sh commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -141,10 +171,56 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 	cache.Lock()
 	defer cache.Unlock()
+
+	if loggerV := cache.logger.V(5); loggerV.Enabled() {
```
It looks useful for me.

pohly replied:
So let's keep it. I added tests for it.

pohly commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -141,10 +171,56 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+	// actively deleted.
+	//
+	// The empty driver name provides the overall count of all active
+	// ResourceClaims regardless of the driver.
```
This is up for debate, see https://github.com/kubernetes/kubernetes/pull/131641#discussion_r2077465713

richabanker commented on test/e2e_node/dra_test.go:
```diff
@@ -382,6 +384,43 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			gomega.Expect(kubeletPlugin1.GetGRPCCalls()).Should(testdriver.NodePrepareResourcesSucceeded, "Plugin 1 should have prepared resources.")
+			gomega.Expect(kubeletPlugin2.GetGRPCCalls()).Should(testdriver.NodePrepareResourcesSucceeded, "Plugin 2 should have prepared resources.")
+			driverName := func(element any) string {
+				el := element.(*model.Sample)
```
Seems like we have allowed to extend the allow list in the past, [ref](https://github.com/kubernetes/kubernetes/issues/89267#issuecomment-1087632859) so I guess we can just do that now? Regarding the usage, I see similar usage of the model package in the codebase to verify the metrics data so should be fine? cc @serathius - the author of the linked issue if he has any ideas on how to avoid importing the package here.

richabanker replied:
Alternatively maybe you could try to create a map representation (map[string]float64) of the vector where the key is the driver_name and the value is the metric value. And then use something like this for verifying the values?

```
claimsInUse := convertVectorToMap(metrics, "dra_resource_claims_in_use")
gomega.Expect(claimsInUse).Should(gstruct.MatchKeys(gstruct.IgnoreExtras, gstruct.Keys{
	"":                   gomega.BeEquivalentTo(1),
	kubeletPlugin1Name: gomega.BeEquivalentTo(1),
	kubeletPlugin2Name: gomega.BeEquivalentTo(1),
}), "metrics while pod is running")
```

pohly replied:
That feels like a workaround. I prefer adding a type alias in `k8s.io/component-base/metrics/testutil`: that package is geared towards use in tests, and already defines a type which directly exposes `model.Sample`, so letting consumers of that package also use that type directly seems fair. The code already depends on it anyway.

In other words, this was possible before:
```
var metrics testutils.Metrics
metrics = ...
samples := metrics["dra_resource_claims_in_use"]
sample := samples[0]
```

It should also be possible to write:
```
var sample *testutil.Sample
sample = samples[0]
```

I suppose some of the importers under https://github.com/kubernetes/kubernetes/blob/3196c9946355c1d20086f66c22e9e5364fb0a56f/hack/verify-prometheus-imports.sh#L72-L79 could use the same approach, but I haven't checked.

For now I have added the type aliases to this PR and use them in `dra_test.go`.

pohly committed:
component-base metrics: allow usage of Prometheus types in tests

Test code using the testutil.Metrics type already depended on the Prometheus
types, but couldn't reference them by name. This is necessary for example when
using Gomega (to cast from `any` in a matcher) or when defining a `var
sample *Sample` which gets set later.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -382,6 +384,43 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			gomega.Expect(kubeletPlugin1.GetGRPCCalls()).Should(testdriver.NodePrepareResourcesSucceeded, "Plugin 1 should have prepared resources.")
+			gomega.Expect(kubeletPlugin2.GetGRPCCalls()).Should(testdriver.NodePrepareResourcesSucceeded, "Plugin 2 should have prepared resources.")
+			driverName := func(element any) string {
+				el := element.(*testutil.Sample)
```
See https://github.com/kubernetes/kubernetes/pull/131641#discussion_r2087385219.

richabanker commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -204,19 +280,81 @@ func (cache *claimInfoCache) syncToCheckpoint() error {
 		}
 	}
-	return cdiDevices
+	counts[""] = total
```
Hmm so Max() also doesnt help here? (Sorry not quite familiar with ResourceClaims and drivers..)
If its unavoidable, maybe include in the HELP for the metric to indicate that sum over this metric is misleading?

Also yeah <any> looks more explicit to me, so +1 to that

pohly replied:
Max is wrong if we have one ResourceClaim for one device from driver A and a second ResourceClaim for one device from driver B. The count for each driver is 1, and thus so is the Max although there are two different ResourceClaims. I'll update the help text.

richabanker commented:
/lgtm

from metrics POV

pohly commented:
@richabanker: can you also approve? It's needed for component-base/metrics.

richabanker commented:
/approve

pohly commented:
@bart0sh: LGTM to you for SIG Node and DRA?

I'll put it into a KEP update before merging, so we will get additional eyes on the proposal later.

bart0sh commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -38,12 +44,18 @@ type ClaimInfo struct {
+// It implements the metrics collector interface.
 type claimInfoCache struct {
+	logger klog.Logger
+	metrics.BaseStableCollector
```
Have you considered making a separate structure for collector? Why do you think embedding is better?

pohly replied:
This seemed simpler. I can move it to a separate struct and then we can compare.

pohly replied:
Pushed. I want to amend commits before we merge, so:
/hold

bart0sh replied:
Up to you. I just thought that it might look a bit clearer if it's separated.

pohly replied:
Does it? I don't have a strong opinion, both seem fine to me.

pohly commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -141,10 +171,56 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+
+// CollectWithStability implements the metrics.StableCollector interface.
+func (cache *claimInfoCache) CollectWithStability(ch chan<- metrics.Metric) {
+	cache.RLock()
```
You mean in the `cache.claimsInUse()` call?

That needs to support being called with the lock already held for the usage in `claimInfoCache.withLock`. I could provide `claimsInUse` and `claimsInUseWithLock`, but I  think letting the caller handle the locking is simpler.

bart0sh replied:
ok, thanks for the explanations, let's keep it as is.

bart0sh commented:
pull-kubernetes-node-e2e-crio-cgrpv1-dra failure seems to be related to this PR.

bart0sh commented:
/triage accepted
/lgtm

@pohly Feel free to unhold after CI failures fixed.

jgehrcke commented:
Thanks! As discussed in a sync call, I wanted to think about this a bit.

So, we want to detect more or less correctly when a driver is "not in use". The proposed method based on Prometheus gauges will provide value towards that.

I think that one fundamental challenge is to reliably distinguish an explicit zero from missing data.

IIUC, we will recommend operators to `sum(dra_resource_claims_in_use{driver_name="some-driver"})`. Effectively, that's a sum over many time series (over all contributing kubelet instances -- running on different nodes).

It's important to know that when one of the contributing series goes away (say, because the node goes down or because a part in the observability pipeline breaks) then in the sum this contribution will drop to zero. This is because of the "staleness" concept in Prometheus (which for gauges typically happens when there hasn't been an update in just a few minutes).

From [Prometheus docs](https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness):

> If a query is evaluated at a sampling timestamp after a time series is marked as stale, then no value is returned for that time series

If I remember correctly, the "no value" in this sentence translates effectively to zero, in a summing operation.

This is
- convenient behavior if the node is gone for good, with all its state (if the old value would keep on living forever that could indeed be an inconvenience)
- non-ideal behavior when the node comes back online later, with previous state

Pragmatic conclusion: maybe we can in the future make sure that we emit an explicit zero at all times. In documentation, we can point out that operators should pay attention to _two_ metrics before concluding that a DRA driver is "not in use":
- `count(dra_resource_claims_in_use{driver_name="some-driver"})` is expected (the node count?)
- `sum(dra_resource_claims_in_use{driver_name="some-driver"})` dropped to zero

jgehrcke commented:
> Is the same true for taking the maximum?

Yes, I am rather certain that a stale time series is treated the same in all of the [aggregate operators](https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators).

(ultimately, we should test if this is true)

As of [this presentation](https://promcon.io/2017-munich/slides/staleness-in-prometheus-2-0.pdf) the "no value" as of staleness is actually a NaN. Looking further... yes: https://github.com/prometheus/prometheus/blob/release-2.29/pkg/value/value.go#L28 -- that's an implementation detail however.

Let's conclude this: if there is no fresh data for a time series for a certain point of time (e.g., "now"), then a design choice in Prometheus is that instead of using an old value they use no value/NaN.

Interesting: a scrape target going away (a scrape failure) is _one_ reason for a time series to go stale even _before_ the five minute interval I mentioned before ([`--query.lookback-delta` config](https://prometheus.io/docs/prometheus/latest/command-line/prometheus/#flags)).

Linking to Brian's talk from a couple of years ago: https://youtu.be/GcTzd2CLH7I?si=qX1sOG7DoxHAw-1o&t=704

In a larger-scale system, it's not unlikely to happen for individual scrapes to fail every now and then -- (a scrape is just an HTTP request that is AFAIR not even retried aggressively).

That is: when guiding operators, we maybe should require them check for stale time series when looking at `dra_resource_claims_in_use`.

How does one do that? I just found this recommendation:

> Generally, instead of looking for stale data, you should alert when the target isn’t being scraped. Prometheus has built in metrics like up for this.

The `up` metric documented [here](https://prometheus.io/docs/concepts/jobs_instances/) is what people typically use for detecting scrape failures.

So, maybe the `up` metric _not_ indicating any scrape problems at the time of evaluating `dra_resource_claims_in_use` is a decent solution.

pohly commented:
It makes sense that aggregation results in NaN. But then why is that not shown to the user, instead reporting 0, which can be confused with a legitimate and (in this case) potentially incorrect value?

Either way, it's obviously a bit more complicated. I'm not sure where to document these best practices. In a page on DRA we shouldn't have to explain how to configure scraping or how to write queries. We cannot even assume that people use Prometheus?!

Do you perhaps want to work on an example dashboard and publish a tutorial or blog post? That then can target Prometheus.
